unblob/handlers/archive/hp/bdl.py

```diff
@@ -0,0 +1,105 @@
+import io
+from pathlib import Path
+from typing import Optional
+
+from dissect.cstruct import Instance
+from structlog import get_logger
+
+from unblob.extractor import carve_chunk_to_file
+from unblob.file_utils import Endian, File, InvalidInputFormat, StructParser, snull
+from unblob.models import Chunk, Extractor, HexString, StructHandler, ValidChunk
+
+logger = get_logger()
+
+C_DEFINITIONS = r"""
+    typedef struct bdl_toc_entry {
+        uint64 offset;
+        uint64 size;
+    } bdl_toc_entry_t;
+
+    typedef struct bdl_header {
+        char magic[4];
+        uint16 major;
+        uint16 minor;
+        uint32 toc_offset;
+        char unknown[4];
+        uint32 toc_entries;
+        uint32 unknowns_2[3];
+        char release[256];
+        char brand[256];
+        char device_id[256];
+        char unknown_3[9];
+        char version[256];
+        char revision[256];
+    } bdl_header_t;
+"""
+
+
+def is_valid_header(header: Instance) -> bool:
+    if header.toc_offset == 0 or header.toc_entries == 0:
+        return False
+    if header.major != 0x1:
+        return False
+    if header.minor != 0x1:
+        return False
+    try:
+        snull(header.release).decode("utf-8")
+        snull(header.brand).decode("utf-8")
+        snull(header.device_id).decode("utf-8")
+        snull(header.version).decode("utf-8")
+        snull(header.revision).decode("utf-8")
+    except UnicodeDecodeError:
+        return False
+    return True
+
+
+class HPBDLExtractor(Extractor):
+    def __init__(self):
+        self._struct_parser = StructParser(C_DEFINITIONS)
+
+    def extract(self, inpath: Path, outdir: Path):
+        with File.from_path(inpath) as file:
+            header = self._struct_parser.parse("bdl_header_t", file, Endian.LITTLE)
+            file.seek(header.toc_offset, io.SEEK_SET)
+            for i in range(header.toc_entries):
+                entry = self._struct_parser.parse(
+                    "bdl_toc_entry_t", file, Endian.LITTLE
+                )
+                next_header_offset = file.tell()
+                carve_chunk_to_file(
+                    file=file,
+                    chunk=Chunk(
+                        start_offset=entry.offset, end_offset=entry.offset + entry.size
+                    ),
+                    carve_path=outdir.joinpath(outdir.joinpath(Path(f"ipkg{i}"))),
+                )
+                file.seek(next_header_offset)
```

Comment:
I would rather collect the start_offset, size instead of seeking back-and forth in the file

Reply:
I was already doing that [here](https://github.com/onekey-sec/unblob/commit/66eda14d4dc60a41afd800dbc8faab7367e8c737) but Quentin said it was possible to do it in a single loop, which recommendation should I follow?